### PR TITLE
Hookes Law implementation for 1D, 2D (plane stress and plane strain) and 3D

### PR DESCRIPTION
**Pull request by alexander-maier**

### CODE REVIEW

File: src/pylife/materiallaws/hookeslaw.py [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+        e12 : array-like float
+            The resulting elastic shear strain component with basis 1-2, 
+            (1. / 2 * e12 is the tensor component)
+        '''
+        s11, s22, s12 = self._inputasarray(s11, s22, s12)
+        self._validateshape(s11, s22, s12)
+        e11 = 1. / self._Et * (s11 - self._nut * s22)
+        e22 = 1. / self._Et * (s22 - self._nut * s11)
+        e33 = - self._nu / self._E * (s11 + s22)
+        e12 = 1. / self._G * s12
```
Comment (johannes-mueller) [inline src/pylife/materiallaws/hookeslaw.py:219]: That would be the enginering shear strain γ12.

**Review by johannes-mueller — CHANGES_REQUESTED** [review]
I have a couple of remarks (see my comments for specific code lines)

There is one general thing that we should clarify prior to merging this.

`e12` would imply that it is the *tensorial* shear strain usually referred to as ε12. However, the equations to calculate it are according to my quick review valid for the *engineering* shear strain, usually referred to as γ12. This would mean that there is a factor 2 missing as γ12 = 2*ε12.

Maybe it's only as matter of definition but we should be really precise here and exactly document what we mean just in order to avoid confusion.

Sources I checked:
https://en.wikipedia.org/wiki/Hooke's_law#Isotropic_materials
https://en.wikipedia.org/wiki/Deformation_(physics)#Shear_strain
https://www.efunda.com/formulae/solid_mechanics/mat_mechanics/hooke_plane_stress.cfm
https://www.efunda.com/formulae/solid_mechanics/mat_mechanics/strain.cfm#engstrain

cc: @DKreuter

**Comment by alexander-maier** [comment]
Thank you for your review.

I agree with you that we should be precise and make clear that `e12` is not the tensor component here. I tried to document this by stating that `1. / 2 * e12 is the tensor component`.

To make things more explicit and avoid confusion I propose to change 
```python
"""
e12 : array-like float
        The resulting elastic shear strain component with basis 1-2, 
        (1. / 2 * e12 is the tensor component)
"""
```
to 
```python
"""
g12 : array-like float
        The resulting elastic engineering shear strain component with basis 1-2, 
        (1. / 2 * g12 is the tensor component)
"""
```

By using a different symbol it should be clear that we are not talking about the tensor component. A disadvantage of this is that we introduce a new symbol, where `g11` is not defined. Regardless of the resolution, we should add the naming convention regarding strain components to [variable_names.rst](https://github.com/boschresearch/pylife/blob/ff72c22efd3ccb75470175330452646ef31f749c/docs/variable_names.rst) in my opinion.

**Comment by johannes-mueller** [comment]
> I agree with you that we should be precise and make clear that `e12` is not the tensor component here. I tried to document this by stating that `1. / 2 * e12 is the tensor component`.

Ah, yes. That's good. In the meantime I have made another test. I took an arbitrary node from `tests/vmap/testfiles/beam_3d_wedge_lin.vmap` and used `Hookeslaw3D` to calculate the strain tensor from the stress tensor and vice versa to compare the results with the FEM results. (see results below). It seems that at least Abaqus also gives out the technical shear strain values and names them `E12`/`E13`/`E23`.

```
In [9]: stress
Out[9]: 
S11    -8.773436
S22     0.581288
S33     0.540051
S12   -22.266996
S13    -2.317505
S23    -9.733004
Name: (3, 8), dtype: float64

In [10]: hook.stress(*strain)
Out[10]: 
(-8.773436304377622,
 0.5812879558039988,
 0.5400506430941474,
 -22.26699593940942,
 -2.3175046283728196,
 -9.733005000247028)

In [11]: strain
Out[11]: 
E11   -4.554919e-11
E22    1.525652e-11
E33    1.498848e-11
E12   -2.894709e-10
E13   -3.012756e-11
E23   -1.265291e-10
Name: (3, 8), dtype: float64

In [12]: hook.strain(*stress)
Out[12]: 
(-4.554918640851974e-11,
 1.5256516814231872e-11,
 1.498847633600235e-11,
 -2.8947095298767094e-10,
 -3.012756037712097e-11,
 -1.265290470123291e-10)
```

File: src/pylife/materiallaws/__init__.py [hunk 2]
```diff
@@ -27,5 +28,9 @@
     'true_strain',
     'true_fracture_stress',
     'true_fracture_strain',
-    'WoehlerCurve'
+    'WoehlerCurve',
+    'Hookeslaw1D',
+    'Hookeslaw2Dstrain',
+    'Hookeslaw2Dstress',
+    'Hookeslaw3D'
```
Comment (johannes-mueller) [inline src/pylife/materiallaws/__init__.py:35]: Generally I would suggest to rename the classes to `HookesLaw*` with capital `L`. Imagine we'd implement some law invented by a guy named "Cole" :smile:

**Comment by johannes-mueller** [comment]
That might be worth a tought. We will probably write an accessor class for the `HooksLaw` classes so that they can directly operate on `PylifeSignal` FEM meshes. In a `StressSignal` the stress components come indeed with capital S. There is no `StrainSignal` accessor class as of yet, but from VMAP we also get the strain variables with capital `E`. So consistency would be nice.

But I wouldn't take it too hard. In `equistress.py` we also have `sxx` with small 's' as parameter names.

**Review by johannes-mueller — APPROVED** [review]
Thanks for the changes.

**Pull request merged**